### PR TITLE
THRIFT-5337 Go set fields write improvement

### PR DESCRIPTION
**Pull request by BytedanceRPC**
There is a duplicate elements check for set in `writeFields*` function,  and it compares elements using `reflect.DeepEqual` which is expensive.

It's much faster that generates a "DeepEqual" function for set elements and apply it in duplicate elements check, especially for nested struct element.

- [x] Did you create an [Apache Jira](https://issues.apache.org/jira/projects/THRIFT/issues/) ticket?  (not required for trivial changes)
- [x] If a ticket exists: Does your pull request title follow the pattern "THRIFT-NNNN: describe my issue"?
- [x] Did you squash your changes to a single commit?  (not required, but preferred)
- [x] Did you do your best to avoid breaking changes?  If one was needed, did you label the Jira ticket with "Breaking-Change"?
- [x] If your change does not involve any code, include `[skip ci]` anywhere in the commit message to free up build resources.

### CODE REVIEW

File: compiler/cpp/src/thrift/generate/t_go_generator.cc [hunk 1]
```diff
@@ -152,6 +152,10 @@ class t_go_generator : public t_generator {
                                  const string& tstruct_name,
                                  bool is_result = false,
                                  bool uses_countsetfields = false);
+  void generate_go_struct_deepequal(std::ostream& out,
```
Comment (fishy) [inline compiler/cpp/src/thrift/generate/t_go_generator.cc:155]: Can we rename all `*deepequal*` into `*equals*`? I don't think it's accurate to call them "deepequal".

File: compiler/cpp/src/thrift/generate/t_go_generator.cc [hunk 2]
```diff
@@ -1847,6 +1860,56 @@ void t_go_generator::generate_go_struct_writer(ostream& out,
   }
 }
 
+void t_go_generator::generate_go_struct_deepequal(ostream& out,
+                                               t_struct* tstruct,
```
Comment (fishy) [inline compiler/cpp/src/thrift/generate/t_go_generator.cc:1864]: nit: I'm not sure what's the coding style for compiler c++ code but this indentation looks weird.

File: compiler/cpp/src/thrift/generate/t_go_generator.cc [hunk 3]
```diff
@@ -1847,6 +1860,56 @@ void t_go_generator::generate_go_struct_writer(ostream& out,
...
+  string name(tstruct->get_name());
+  const vector<t_field*>& fields = tstruct->get_sorted_members();
+  vector<t_field*>::const_iterator f_iter;
+  indent(out) << "func (p *" << tstruct_name << ") " << deepequal_method_name_ << "(other *"
+              << tstruct_name << ") bool {" << endl;
+  indent_up();
+
+  string field_name;
+  string publicize_field_name;
+  int32_t field_id = -1;
```
Comment (fishy) [inline compiler/cpp/src/thrift/generate/t_go_generator.cc:1879]: this is unused and it will cause certain compilers to fail:

```
src/thrift/generate/t_go_generator.cc:1879:11: error: unused variable ‘field_id’ [-Werror=unused-variable]
 1879 |   int32_t field_id = -1;
      |           ^~~~~~~~
cc1plus: all warnings being treated as errors
```

**Review by fishy — COMMENTED** [review]
I haven't gone through all the changes yet, but wanted to give you some early feedback.

File: compiler/cpp/src/thrift/generate/t_go_generator.cc [hunk 4]
```diff
@@ -1847,6 +1860,56 @@ void t_go_generator::generate_go_struct_writer(ostream& out,
...
+    publicize_field_name = publicize(field_name);
+    string goType = type_to_go_type_with_opt(field_type, is_pointer_field(*f_iter));
+
+    string tgt = "p." + publicize_field_name;
+    string src = "other." + publicize_field_name;
+    t_type* ttype = field_type->get_true_type();
+    if (is_pointer_field(*f_iter)
+        && (ttype->is_base_type() || ttype->is_enum() || ttype->is_container())) {
+    tgt = "(*" + tgt + ")";
+    src = "(*" + src + ")";
```
Comment (fishy) [inline compiler/cpp/src/thrift/generate/t_go_generator.cc:1902]: 1. wrong indentation
2. This generates code like this for optional fields:
```go
if (*p.FieldN) != (*other.FieldN) { return false }
```
this will cause panic when the field is not set.

File: compiler/cpp/src/thrift/generate/t_go_generator.cc (hunk 1 above)
Comment (BytedanceRPC) [reply compiler/cpp/src/thrift/generate/t_go_generator.cc]: Thanks for review, *equals* sounds better.

File: compiler/cpp/src/thrift/generate/t_go_generator.cc (hunk 2 above)
Comment (BytedanceRPC) [reply compiler/cpp/src/thrift/generate/t_go_generator.cc:1866]: Well, I use clang-format according to https://github.com/apache/thrift/blob/master/.clang-format

File: compiler/cpp/src/thrift/generate/t_go_generator.cc (hunk 3 above)
Comment (BytedanceRPC) [reply compiler/cpp/src/thrift/generate/t_go_generator.cc]: removed latest commit

File: compiler/cpp/src/thrift/generate/t_go_generator.cc (hunk 4 above)
Comment (BytedanceRPC) [reply compiler/cpp/src/thrift/generate/t_go_generator.cc]: Thanks, I add a pointer check to avoid panic last commit

**Comment by dcelasun** [comment]
Looks reasonable at a first glance, will do a proper review later.

I wonder if there are any cases covered by `reflect.DeepEqual` that aren't covered by this method. If there are, that would be a BC break and we'd have to mention it in the changelog.

**Comment by BytedanceRPC** [comment]
@dcelasun The process of `Equals` is pretty like `Write` (deserialize). Since `write` can cover all cases, I believe so can `Equals`.

File: compiler/cpp/src/thrift/generate/t_go_generator.cc [hunk 5]
```diff
@@ -1847,6 +1862,48 @@ void t_go_generator::generate_go_struct_writer(ostream& out,
...
+    field_name = (*f_iter)->get_name();
+    t_type* field_type = (*f_iter)->get_type();
+    publicize_field_name = publicize(field_name);
+    string goType = type_to_go_type_with_opt(field_type, is_pointer_field(*f_iter));
+
+    string tgt = "p." + publicize_field_name;
+    string src = "other." + publicize_field_name;
+    t_type* ttype = field_type->get_true_type();
+    if (is_pointer_field(*f_iter)
+        && (ttype->is_base_type() || ttype->is_enum() || ttype->is_container())) {
```
Comment (fishy) [inline compiler/cpp/src/thrift/generate/t_go_generator.cc:1893]: I'm not sure why the condition after `&&` is needed, can you explain?

File: compiler/cpp/src/thrift/generate/t_go_generator.cc [hunk 6]
```diff
@@ -3439,6 +3511,123 @@ void t_go_generator::generate_serialize_list_element(ostream& out, t_list* tlist
...
+                                                  string src) {
+  (void)ttype;
+  out << indent() << "if !" << tgt << "." << equals_method_name_ << "(" << src
+      << ") { return false }" << endl;
+}
+
+/**
+ * Compares any container type
+ */
+void t_go_generator::generate_go_equals_container(std::ostream& out,
```
Comment (fishy) [inline compiler/cpp/src/thrift/generate/t_go_generator.cc:3599]: For containers, another optimization you can do is to compare the sizes first. If the sizes doesn't match you can return `false` directly.

**Review by fishy — COMMENTED** [review]
I like where this is going, but left some small comments.

Also I believe with this change there's no other way the `reflect` import is used, so we can totally remove it from compiled code now.

**Comment by fishy** [comment]
I believe currently there's some bug in your implementation. The result is different from `reflect.DeepEqual`, Here's the code I used:

`test.thrift`:
```thrift
namespace go test

enum Enum {
  One = 1,
  Two = 2,
}

struct Struct {
  1: optional string Str,
  2: optional i32 I32,
  3: optional set<Enum> Enums,
}
```

`bench_test.go`:
```go
package thriftsetbench

import (
        "reflect"
        "testing"

        "github.com/apache/thrift/lib/go/thrift"
        "github.com/fishy/thriftsetbench/gen-go/test"
)

func BenchmarkEqual(b *testing.B) {
        equalGetter := func() (*test.Struct, *test.Struct) {
                s1 := &test.Struct{
                        Str: thrift.StringPtr("foobar"),
                        Enums: []test.Enum{
                                test.Enum_One,
                                test.Enum_Two,
                        },
                }
                s2 := &test.Struct{
                        Str: thrift.StringPtr("foobar"),
                        Enums: []test.Enum{
                                test.Enum_One,
                                test.Enum_Two,
                        },
                }
                return s1, s2
        }
        nonEqualGetter := func() (*test.Struct, *test.Struct) {
                s1 := &test.Struct{
                        Str: thrift.StringPtr("foobar"),
                        Enums: []test.Enum{
                                test.Enum_One,
                        },
                }
                s2 := &test.Struct{
                        Str: thrift.StringPtr("foobar"),
                        Enums: []test.Enum{
                                test.Enum_Two,
                        },
                }
                return s1, s2
        }

        for label, getter := range map[string]func() (*test.Struct, *test.Struct){
                "equal":    equalGetter,
                "nonequal": nonEqualGetter,
        } {
                b.Run(label, func(b *testing.B) {
                        s1, s2 := getter()
                        reflectResult := reflect.DeepEqual(s1, s2)
                        equalsResult := s1.Equals(s2)
                        if reflectResult != equalsResult {
                                b.Errorf("Reflect result: %v, Equals result: %v", reflectResult, equalsResult)
                        }

                        b.Run("reflect", func(b *testing.B) {
                                b.RunParallel(func(pb *testing.PB) {
                                        for pb.Next() {
                                                reflect.DeepEqual(s1, s2)
                                        }
                                })
                        })

                        b.Run("equals", func(b *testing.B) {
                                b.RunParallel(func(pb *testing.PB) {
                                        for pb.Next() {
                                                s1.Equals(s2)
                                        }
                                })
                        })
                })
        }
}
```

result:
```
$ go test -bench . -benchmem
goos: linux
goarch: amd64
pkg: github.com/fishy/thriftsetbench
BenchmarkEqual/equal/reflect-12         10597189               113 ns/op              64 B/op          6 allocs/op
BenchmarkEqual/equal/equals-12          1000000000               0.680 ns/op           0 B/op          0 allocs/op
BenchmarkEqual/nonequal/reflect-12              12201334                99.9 ns/op            48 B/op          4 allocs/op
BenchmarkEqual/nonequal/equals-12               1000000000               0.712 ns/op           0 B/op          0 allocs/op
--- FAIL: BenchmarkEqual/nonequal
    bench_test.go:54: Reflect result: false, Equals result: true
--- FAIL: BenchmarkEqual
FAIL
exit status 1
FAIL    github.com/fishy/thriftsetbench 4.183s
```

We likely would need to have some thorough test for the `Equals` implementation with this PR.

File: compiler/cpp/src/thrift/generate/t_go_generator.cc (hunk 5 above)
Comment (BytedanceRPC) [reply compiler/cpp/src/thrift/generate/t_go_generator.cc:1900]: for struct and xception, since they have `Equals` method, there is no need to get their value

File: compiler/cpp/src/thrift/generate/t_go_generator.cc (hunk 6 above)
Comment (BytedanceRPC) [reply compiler/cpp/src/thrift/generate/t_go_generator.cc:3602]: Good idea, will do

**Comment by BytedanceRPC** [comment]
@fishy It's a bug in `generate_go_equals_guard`, I fixed it. 
Yes, some thorough test is pretty necessary. I will also do more test.

**Comment by fishy** [comment]
@BytedanceRPC there's another go compiler change merged, so can you rebase your PR on latest master branch?

Also please add some tests for the `Equals` implementation. This change looks promising, but we need the confidence that it's implemented correctly.

**Comment by BytedanceRPC** [comment]
@fishy rebased PR on latest master branch, and added `Equals` test under lib/go/test

File: compiler/cpp/src/thrift/generate/t_go_generator.cc [hunk 7]
```diff
@@ -1851,6 +1864,65 @@ void t_go_generator::generate_go_struct_writer(ostream& out,
   }
 }
 
+void t_go_generator::generate_go_struct_equals(ostream& out,
+                                               t_struct* tstruct,
+                                               const string& tstruct_name,
+                                               bool is_args_or_result) {
+  if (is_args_or_result) {
+    return;
+  }
```
Comment (fishy) [inline compiler/cpp/src/thrift/generate/t_go_generator.cc:1873]: nit: I think we should do this check on the call site instead, and currently there's only one caller of it (Line 1497), so there should be:

```
if !is_result && !is_args {
  generate_go_struct_equals(out, tstruct, tstruct_name);
}
```

Semantically this function is called "generate_*", but when one of its arg is passed in true it doesn't generate anything, which feels wrong to me.

File: compiler/cpp/src/thrift/generate/t_go_generator.cc [hunk 8]
```diff
@@ -3463,6 +3550,112 @@ void t_go_generator::generate_serialize_list_element(ostream& out, t_list* tlist
   generate_serialize_field(out, &efield, prefix);
 }
 
+/**
+ * Compares any type
+ */
+void t_go_generator::generate_go_equals(std::ostream& out,
+                                           t_type* ori_type,
```
Comment (fishy) [inline compiler/cpp/src/thrift/generate/t_go_generator.cc:3557]: nit: this indentation looks weird. usually you either align with `(`, or use fixed 2/4/6 spaces. this is neither.

I don't believe clang-format is used to generate this indentation either, at least it's not enforced by travis. when I run `clang-format -i --style=file compiler/cpp/src/thrift/generate/t_go_generator.cc` it made a ton of changes in this file.

File: lib/go/test/EqualsTest.thrift [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  10: optional i64 OptI64Foo,
+  11: double DoubleFoo,
+  12: optional double OptDoubleFoo,
+  13: string StrFoo,
+  14: optional string OptStrFoo,
+  15: binary BinFoo,
+  16: optional binary OptBinFoo,
+  17: EnumFoo EnumFoo,
+  18: optional EnumFoo OptEnumFoo,
+}
```
Comment (fishy) [inline lib/go/test/EqualsTest.thrift:25]: I would suggest also add a few typedef'd primitive types into this (and also into `ListEqualsFoo`, `SetEqualsFoo`, `MapEqualsFoo`).

File: lib/go/test/tests/equals_test.go [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+import (
+	"equalstest"
+	"strconv"
+	"testing"
+)
+
+func TestEquals(t *testing.T) {
+	basicTgt, basicSrc := genBasicFoo(), genBasicFoo()
+	if !basicTgt.Equals(basicSrc) {
+		t.Fatal("BasicEqualsFoo.Equals() test failed")
```
Comment (fishy) [inline lib/go/test/tests/equals_test.go:12]: These test failures should be `t.Error` instead of `t.Fatal`. Fatal stops the test from running, so if someone made a mistake in the future and failed multiple cases inside this test, they will only see the first failure in the output.

Fatal should only be used when the following tests are depending on this one (e.g. used for nil checks).

File: lib/go/test/tests/equals_test.go [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+	}
+	mapTgt, mapSrc := genMapFoo(), genMapFoo()
+	if !mapTgt.Equals(mapSrc) {
+		t.Fatal("MapEqualsFoo.Equals() test failed")
+	}
+	mapSrc.OptI64I64StringMapMapFoo[1][1] = "0"
+	if mapTgt.Equals(mapSrc) {
+		t.Fatal("MapEqualsFoo.Equals() test failed")
+	}
+}
```
Comment (fishy) [inline lib/go/test/tests/equals_test.go:50]: Another issue with the test is that all the optional fields are set, so if someone missed the nil check in the compiler (like in the first version of this PR) this test would not catch that error. I would suggest add a few more cases to each `gen*` call, while changing some of the optional fields to `nil` (for example, you should add a test to change one of the optional field to nil and make sure `Equals` return false, then change the other also to nil and make sure `Equals` return true).

**Review by fishy — COMMENTED** [review]
Overall lgtm, but there are a few things can be improved in the test code.

Also you can remove https://github.com/apache/thrift/blob/c843abbe20427f07143266ca565a820945ff8d23/compiler/cpp/src/thrift/generate/t_go_generator.cc#L927 and https://github.com/apache/thrift/blob/c843abbe20427f07143266ca565a820945ff8d23/compiler/cpp/src/thrift/generate/t_go_generator.cc#L952. Since this change eliminated the usage of `reflect.DeepEqual`, the reflect package is not used any more in compiled go code.

File: compiler/cpp/src/thrift/generate/t_go_generator.cc (hunk 7 above)
Comment (BytedanceRPC) [reply compiler/cpp/src/thrift/generate/t_go_generator.cc:1873]: looks reasonable, thanks for suggestion.

File: compiler/cpp/src/thrift/generate/t_go_generator.cc (hunk 8 above)
Comment (BytedanceRPC) [reply compiler/cpp/src/thrift/generate/t_go_generator.cc:3557]: not sure is the `std::` making indentation looks weird. Indentations looks proper after I remove `std::` and apply `clang-format`.
To avoid unnecessary change, I only apply the output of clang-format on my changes in `compiler/cpp/src/thrift/generate/t_go_generator.cc`.

File: lib/go/test/EqualsTest.thrift (hunk 9 above)
Comment (BytedanceRPC) [reply lib/go/test/EqualsTest.thrift:35]: good suggestion

File: lib/go/test/tests/equals_test.go (hunk 10 above)
Comment (BytedanceRPC) [reply lib/go/test/tests/equals_test.go]: Thanks for suggestion

File: lib/go/test/tests/equals_test.go (hunk 11 above)
Comment (BytedanceRPC) [reply lib/go/test/tests/equals_test.go:95]: Yes, I missed `optional field`'s pointer check test.

File: compiler/cpp/src/thrift/generate/t_go_generator.cc [hunk 12]
```diff
@@ -3463,6 +3543,109 @@ void t_go_generator::generate_serialize_list_element(ostream& out, t_list* tlist
...
+}
+
+/**
+ * Compares the members of a struct
+ */
+void t_go_generator::generate_go_equals_struct(ostream& out,
+                                               t_type* ttype,
+                                               string tgt,
+                                               string src) {
+  (void)ttype;
```
Comment (fishy) [inline compiler/cpp/src/thrift/generate/t_go_generator.cc:3608]: what does this line do?

File: compiler/cpp/src/thrift/generate/t_go_generator.cc [hunk 13]
```diff
@@ -3463,6 +3543,109 @@ void t_go_generator::generate_serialize_list_element(ostream& out, t_list* tlist
...
+  if (ttype->is_map()) {
+    t_map* tmap = (t_map*)ttype;
+    out << indent() << "for k, _tgt := range " << tgt << " {" << endl;
+    indent_up();
+    string element_source = tmp("_src");
+    out << indent() << element_source << " := " << src << "[k]" << endl;
+    generate_go_equals(out, tmap->get_val_type(), "_tgt", element_source);
+    indent_down();
+    indent(out) << "}" << endl;
+  } else if (ttype->is_list() || ttype->is_set()) {
```
Comment (fishy) [inline compiler/cpp/src/thrift/generate/t_go_generator.cc:3630]: should this just be `} else {`, or maybe add another `} else { throw "unsupported container" }` at the end? currently if this is a container but it's not a map/list/set (I'm not sure if we have other container types) this function only checks for length and then does nothing.

File: lib/go/test/tests/equals_test.go [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+		OptI64Foo:    &(&struct{ x int64 }{4}).x,
+		DoubleFoo:    5,
+		OptDoubleFoo: &(&struct{ x float64 }{5}).x,
+		StrFoo:       "6",
+		OptStrFoo:    &(&struct{ x string }{"6"}).x,
+		BinFoo:       []byte("7"),
+		OptBinFoo:    []byte("7"),
+		EnumFoo:      equalstest.EnumFoo_e1,
+		OptEnumFoo:   &(&struct{ x equalstest.EnumFoo }{equalstest.EnumFoo_e1}).x,
+	}
```
Comment (fishy) [inline lib/go/test/tests/equals_test.go:117]: the newly added typedef'd fields are not added here (and other `gen*` functions)

File: test/go/src/bin/testequals/main.go [hunk 15]
```diff
@@ -0,0 +1,179 @@
+package main
```
Comment (fishy) [inline test/go/src/bin/testequals/main.go:1]: hmm why do we need to add this directory/file?

Comment (BytedanceRPC) [reply test/go/src/bin/testequals/main.go:1]: sry, wrong push

File: compiler/cpp/src/thrift/generate/t_go_generator.cc (hunk 12 above)
Comment (BytedanceRPC) [reply compiler/cpp/src/thrift/generate/t_go_generator.cc:3608]: try to keep `generate_go_equals*` with the same arguments, but `ttype` not really used here.

File: compiler/cpp/src/thrift/generate/t_go_generator.cc (hunk 13 above)
Comment (BytedanceRPC) [reply compiler/cpp/src/thrift/generate/t_go_generator.cc:3630]: throw exception in else condition should work well according to `generate_deserialize_container`

File: lib/go/test/tests/equals_test.go (hunk 14 above)
Comment (BytedanceRPC) [reply lib/go/test/tests/equals_test.go:123]: sry, wrong push

**Review by fishy — APPROVED** [review]
Thanks. Looks good. I'll merge it once the tests pass.

File: lib/go/test/tests/equals_test.go [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+		BinFoo:       []byte("7"),
+		OptBinFoo:    []byte("7"),
+		EnumFoo:      equalstest.EnumFoo_e1,
+		OptEnumFoo:   &(&struct{ x equalstest.EnumFoo }{equalstest.EnumFoo_e1}).x,
+		MyByteFoo:    equalstest.Mybyte(8),
+		OptMyByteFoo: &(&struct{ x equalstest.Mybyte }{8}).x,
+		MyStrFoo:     equalstest.Mystr("9"),
+		OptMyStrFoo:  &(&struct{ x equalstest.Mystr }{"9"}).x,
+		MyBinFoo:     equalstest.Mybin("10"),
+		OptMyBinFoo:  &(&struct{ x equalstest.Mybin }{equalstest.Mybin("10")}).x,
```
Comment (fishy) [inline lib/go/test/tests/equals_test.go:122]: This one does not compile, see https://travis-ci.org/github/apache/thrift/jobs/757426362

**Review by fishy — CHANGES_REQUESTED** [review]

File: lib/go/test/tests/equals_test.go (hunk 16 above)
Comment (fishy) [reply lib/go/test/tests/equals_test.go:122]: Oh I think you just misses another level of `&()` around the whole thing.

But this is getting really hard to read so please just use some local variables instead.

Comment (fishy) [reply lib/go/test/tests/equals_test.go:122]: You can just write them this way, which is way simpler:

```go
                OptEnumFoo:   equalstest.EnumFooPtr(equalstest.EnumFoo_e1),
                MyByteFoo:    equalstest.Mybyte(8),
                OptMyByteFoo: equalstest.MybytePtr(8),
                MyStrFoo:     equalstest.Mystr("9"),
                OptMyStrFoo:  equalstest.Mystr("9"),
                MyBinFoo:     equalstest.Mybin("10"),
                OptMyBinFoo:  equalstest.MybinPtr(equalstest.Mybin("10")),
```

File: lib/go/test/tests/equals_test.go [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+		StrFoo:       "6",
+		OptStrFoo:    &(&struct{ x string }{"6"}).x,
+		BinFoo:       []byte("7"),
+		OptBinFoo:    []byte("7"),
+		EnumFoo:      equalstest.EnumFoo_e1,
+		OptEnumFoo:   &(&struct{ x equalstest.EnumFoo }{equalstest.EnumFoo_e1}).x,
+		MyByteFoo:    equalstest.Mybyte(8),
+		OptMyByteFoo: &(&struct{ x equalstest.Mybyte }{8}).x,
+		MyStrFoo:     equalstest.Mystr("9"),
+		OptMyStrFoo:  &(&struct{ x equalstest.Mystr }{"9"}).x,
```
Comment (fishy) [inline lib/go/test/tests/equals_test.go:120]:
```suggestion
		OptMyStrFoo:  equalstest.Mystr("9"),
```

File: lib/go/test/tests/equals_test.go [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+		DoubleFoo:    5,
+		OptDoubleFoo: &(&struct{ x float64 }{5}).x,
+		StrFoo:       "6",
+		OptStrFoo:    &(&struct{ x string }{"6"}).x,
+		BinFoo:       []byte("7"),
+		OptBinFoo:    []byte("7"),
+		EnumFoo:      equalstest.EnumFoo_e1,
+		OptEnumFoo:   &(&struct{ x equalstest.EnumFoo }{equalstest.EnumFoo_e1}).x,
+		MyByteFoo:    equalstest.Mybyte(8),
+		OptMyByteFoo: &(&struct{ x equalstest.Mybyte }{8}).x,
```
Comment (fishy) [inline lib/go/test/tests/equals_test.go:118]:
```suggestion
		OptMyByteFoo: equalstest.MybytePtr(8),
```

File: lib/go/test/tests/equals_test.go [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+		I64Foo:       4,
+		OptI64Foo:    &(&struct{ x int64 }{4}).x,
+		DoubleFoo:    5,
+		OptDoubleFoo: &(&struct{ x float64 }{5}).x,
+		StrFoo:       "6",
+		OptStrFoo:    &(&struct{ x string }{"6"}).x,
+		BinFoo:       []byte("7"),
+		OptBinFoo:    []byte("7"),
+		EnumFoo:      equalstest.EnumFoo_e1,
+		OptEnumFoo:   &(&struct{ x equalstest.EnumFoo }{equalstest.EnumFoo_e1}).x,
```
Comment (fishy) [inline lib/go/test/tests/equals_test.go:116]:
```suggestion
		OptEnumFoo:   equalstest.EnumFooPtr(equalstest.EnumFoo_e1),
```

File: lib/go/test/tests/equals_test.go (hunk 16 above)
Comment (fishy) [reply lib/go/test/tests/equals_test.go:122]: Oh right it's because for binary type optional are the same as required (because slices are already pointers), so this should be:

```suggestion
		OptMyBinFoo:  equalstest.MybinPtr(equalstest.Mybin("10")),
```

Comment (fishy) [reply lib/go/test/tests/equals_test.go:122]: This change passes the tests: https://travis-ci.org/github/apache/thrift/builds/757454619

https://github.com/apache/thrift/pull/2321/commits/8b820c3dd40a00c7f3243dff496eb59730e2d727 are all the changes I made but I believe only the change on this line is required (but I'd still suggest you to make other changes as well, and run `gofmt -s` on this file)

**Comment by fishy** [comment]
Can you squash your PR to a single commit, with proper commit message (what you used in the PR title and description)?

File: lib/go/test/tests/equals_test.go (hunk 16 above)
Comment (BytedanceRPC) [reply lib/go/test/tests/equals_test.go]: I think it shoud be just `OptMyBinFoo:  equalstest.Mybin("10"),`

**Review by fishy — APPROVED** [review]
Travis should finish in 20min or so and I'll merge it after that.
